File: exam_at_home/1/bracket_matching.cc
/*
* author: wang-yang
* email: tnst92002@gmail.com
*/
#include <bits/stdc++.h>
using namespace std;

bool check(string s) {
    stack <char> st;
    for (char c : s) {
        switch (c) {
            case '(':
                st.emplace(')');
                break;
            case '[':
                st.emplace(']');
                break;
            case ')':
            case ']':
                if (st.empty() || st.top() != c)
                    return false;
                st.pop();
                break;
            default:
                __builtin_unreachable();
        }
    }
    return st.empty();
}

int main() {
    int N;
    cin >> N;
    while (N--) {
        string buf;
        cin >> buf;
        if (check(buf)) cout << "Yes\n";
        else cout << "No\n";
    }
}

File: exam_at_home/1/bracket_matching.py
#
# author: wang-yang
# email: tnst92002@gmail.com
#
N = int(input())

def check(s: str):
    st = []
    for c in s:
        if c == '(':
            st.append(')')
        elif c == '[':
            st.append(']')
        else:
            if len(st) == 0 or st[-1] != c:
                return False
            st.pop()
    return len(st) == 0

for _ in range(N):
    s = input()
    st = []
    print("Yes" if check(s) else "No")